src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -1567,6 +1569,9 @@ private async Task CreateContainerAsync(AppResource cr, ILogger resourceLogger,
         }
 
         await kubernetesService.CreateAsync(dcpContainerResource, cancellationToken).ConfigureAwait(false);
+
+        var containerStartedEvent = new ContainerResourceStartedEvent(modelContainerResource);
+        await eventing.PublishAsync(containerStartedEvent, cancellationToken).ConfigureAwait(false);
```

Comment:
My understanding is that kubernetes follows a "desired state" and "actual state" model. Just because the `CreateAsync` call completed doesn't necessarily mean the container has been started.

So are we going to have timing issues that this "container resource started" event is fired, but the actual container hasn't been started yet?

cc @karolz-ms

Reply:
@eerhardt is right. If by "container started" we want to convey "it is running and there is good chance that it will reply to requests", then one needs to wait for `Container.Status.State` property https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/Model/Container.cs#L259 to become "Running". And in future, an even better event to have would be when the Container becomes healthy.

Reply:
This is just an example event the idea of the pub sub model is that we can fire specific events at lots of different places. In the scenario you've listed above we'd probably listen to the resource notifications coming back.

The reality is that even if we fire the event a bit later there is no guarantee it'll be live for servicing requests.

Reply:
I think I'm confused at why I would use this new `ContainerResourceStartedEvent` capability over listening to the resource notifications. Getting an event before something is ready to use isn't great.

We are hitting these situations all over in our tests - needing to know when something is ready, or stopped. See

https://github.com/dotnet/aspire/blob/0d8b294cf7d3423e9b4a6473b09166f4c46e9c7a/tests/Aspire.Hosting.Testing.Tests/TestingBuilderTests.cs#L67-L70

https://github.com/dotnet/aspire/issues/4878

Are you saying these new, higher-level events would be built on top of the ResourceNotificationService? Like that "WaitForTextAsync" is today?

https://github.com/dotnet/aspire/blob/0d8b294cf7d3423e9b4a6473b09166f4c46e9c7a/tests/Aspire.Hosting.Tests/Utils/LoggerNotificationExtensions.cs#L59-L80

So we'd have 2 levels of "events"? Lower level `ResourceNotificationService`. And then higher-level `IDistributedApplicationEventing`.

Reply:
The way to think of this new API is a more flexible version of lifecycle events which adds the ability to publish and subscribe to arbitrary events without having to implement an interface and inject it into DI.

I would expect that some extensions would use the feed from the resource notification service to fire more specific events related to a particular resource.

Reply:
One big difference with RNS is that the events are synchronous, as in, they run I’ll inline. RNS publishes events and they can’t be waited on inline and that means the publisher can’t react to mutations during the event nor can it wait for the event to complete before progressing the workflow.